pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,382 @@
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	defer q.wg.Done()
+
+	if q.list.Len() == 0 {
+		// the list is empty; no further flushing needs to be triggered
+		q.attempt = 0
+		q.scheduled = false
+		return
+	}
+	// the list is not empty
+	q.scheduled = true
+	if retries > 0 {
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		q.wg.Add(1)
+		q.timer = time.AfterFunc(backoffDuration(q.attempt), q.flush)
+		return
+	}
+	// all items in the list are new; flush immediately
+	q.wg.Add(1)
+	go q.flush()
+}
+
+// drainQueue drains the entire queue and returns all the payloads that were in it.
+func (q *sender) drainQueue() []*payload {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	var payloads []*payload
+	for q.list.Len() > 0 {
+		v := q.list.Remove(q.list.Front())
+		payloads = append(payloads, v.(*payload))
+	}
+	q.size = 0
+	return payloads
+}
+
+// sendPayloads concurrently sends the given list of payloads. It returns
+// the number of payloads that were added back onto the queue to be retried
+// again later due to an error.
+func (q *sender) sendPayloads(payloads []*payload) (done, retries uint64) {
+	var wg sync.WaitGroup
+	for _, p := range payloads {
+		q.climit <- struct{}{}
+		wg.Add(1)
+		go func(p *payload) {
+			defer func() { <-q.climit }()
+			defer wg.Done()
+			req, err := p.httpRequest(q.cfg.url)
+			if err != nil {
+				log.Errorf("http.Request: %s", err)
+				return
+			}
+			start := time.Now()
+			err = q.do(req)
+			stats := &eventData{
+				bytes:    len(p.body),
+				count:    1,
+				duration: time.Since(start),
+				err:      err,
+			}
+			switch err.(type) {
+			case *retriableError:
+				// request failed again, but can be retried
+				q.enqueue(p)
+				atomic.AddUint64(&retries, 1)
+				q.recordEvent(eventTypeRetry, stats)
+			case nil:
+				// request was successful
+				atomic.AddUint64(&done, 1)
+				q.recordEvent(eventTypeSent, stats)
+			default:
+				// this is a fatal error, we have to drop this payload
+				q.recordEvent(eventTypeFailed, stats)
+			}
+		}(p)
+	}
+	wg.Wait()
+	return
+}
+
+// recordEvent records that event t has happened and attaches it the given data.
+func (q *sender) recordEvent(t eventType, data *eventData) {
+	if recorder := q.cfg.recorder; recorder != nil {
+		data.host = q.cfg.url.Hostname()
+		recorder.recordEvent(t, data)
+	}
+}
+
+// Flush waits up to 5 seconds for the queue to reach an empty state and for all scheduling
+// to complete before returning.
+func (q *sender) Flush() {
```

Comment:
I don't think this should be part of the public interface of a sender since it gives the false impression that payloads are buffered and flushed at arbitrary points in time when in fact flush is attempted after each payload. Also, the fact that no return feedback is given makes it kinda weird as you're never sure if something actually flushed or not.

I think this is mostly being used for tests and for stopping logic. Would it make sense to just call this `Stop`/`Close, wait for empty pipes, free resources and prevent new stuff from being enqueued as you as soon as you called it?

Reply:
I think this method is honestly the one I've been stuck on the most. I'm not sure what the best approach is. I think that adding more boolean flags throughout of the form `if q.stopped { return }` is not necessary and just makes everything a bit yuckier (I did try it). Furthermore, there is nothing to actually stop.

The logic here was that we try and wait for a bit for the queue to be emptied (potentially doing some more retries) and then return.

I do agree that the name is confusing. But I don't have any good ideas: `stopping`, `waitEmpty`? I'll try `waitEmpty` but leave this comment unresolved for any better ideas.

Reply:
My issue was mostly with it being exposed as part of the "public" interface. I'm fine with its current state as `waitEmpty` specially if there's nothing to stop.